krm-functions/lib/kubeobject/kubeobject.go

```diff
@@ -79,9 +79,11 @@ func (o *KubeObjectExt[T1]) SetStatus(newStatus interface{}) error {
 	return safeSetNestedFieldKeepFormatting(&o.KubeObject, newStatus, "status")
 }
 
-// NOTE: the following functions are considered as "methods" of KubeObject,
-// and thus nill checking of `obj` was omitted intentionally:
-// the caller is responsible for ensuring that `obj` is not nil`
+// SetFromTypedObject sets the value of `o` to `value`, while keeping most of the YAML formatting.
+// It can be seen as an in-place version of fn.NewFromTypedObject
+func (o *KubeObjectExt[T1]) SetFromTypedObject(value T1) error {
```

Comment:
When testing this this does not work. When applying SetStatus it works fine.

func (r *fnCtx) updateIPAllocationResource(forObj *fn.KubeObject, objs fn.KubeObjects) (*fn.KubeObject, error) {
if forObj == nil {
return nil, fmt.Errorf("expected a for object but got nil")
}
fn.Logf("ipalloc: %v\n", forObj)
allocKOE, err := ko.NewFromKubeObject[*ipamv1alpha1.IPAllocation](https://github.com/nephio-project/nephio/pull/142/forObj)
if err != nil {
return nil, err
}
alloc, err := allocKOE.GetGoStruct()
if err != nil {
return nil, err
}
resp, err := r.IpamClientProxy.Allocate(context.Background(), alloc, nil)
if err != nil {
return nil, err
}
alloc.Status = resp.Status

if alloc.Status.Prefix != nil {
	fn.Logf("ipalloc resp prefix: %v\n", *resp.Status.Prefix)
}
if alloc.Status.Gateway != nil {
	fn.Logf("ipalloc resp gateway: %v\n", *resp.Status.Gateway)
}
// set the status
//err = allocKOE.SetStatus(resp.Status)
err = allocKOE.SetFromTypedObject(resp)
return &allocKOE.KubeObject, err
}

testable using github.com/henderiw-nephio/pkg-examples -> ipam-fn

Reply:
What is the error message that you got?

For me the following testcase is green: (however see my next comment)

```
func TestIpalloc(t *testing.T) {
	var inputYaml string = `
apiVersion: ipam.alloc.nephio.org/v1alpha1
kind: IPAllocation
metadata:
  name: internetpool1
spec:
  kind: pool
  selector:
  matchLabels:
    nephio.org/site: edge1
  networkInstance: {}
  prefixLength: 8
`

	allocKOE, err := NewFromYaml[*ipamv1alpha1.IPAllocation]([]byte(inputYaml))
	assertNoError(t, err)
	alloc, err := allocKOE.GetGoStruct()
	assertNoError(t, err)

	alloc.Status.Gateway = "2.2.2.2"
	alloc.Status.AllocatedPrefix = "3.3.3.0"

	// set the status
	//err = allocKOE.SetStatus(resp.Status)
	err = allocKOE.SetFromTypedObject(alloc)
	assertNoError(t, err)

	var expectedYaml string = `apiVersion: ipam.alloc.nephio.org/v1alpha1
kind: IPAllocation
metadata:
  name: internetpool1
spec:
  kind: pool
  networkInstance: {}
  prefixLength: 8
status:
  prefix: 3.3.3.0
  gateway: 2.2.2.2`

	actualYaml := strings.TrimSpace(allocKOE.String())
	if actualYaml != expectedYaml {
		t.Errorf(actualYaml)
	}

}
```

Reply:
Having said that you are using `KubeObjectExt[*ipamv1alpha1.IPAllocation]`.
So far I thought that KubeObjectExt's type parameter should be a struct , as opposed to a pointer-to-struct. I came to this conclusion by lookinto into kubeobject_test.go (even the parts that wasn't written by me :)), or into the implementation of `GetGoStruct()`:
GetGoStruct accidentally works with the pointer type parameter, but in that case we are passing a `**ipamv1alpha1.IPAllocation` to `As()`, while it was clearly designed to receive a `*ipamv1alpha1.IPAllocation`.

This might cause your different behaviour between SetFromTypedObject and SetSpec. For SetSPec you pass `resp.Status` that is a struct, but to SetFromTypedObject() you pass `resp` that is a pointer-to-struct.

The important question here: Should we support pointer type parameters, as well?

Reply:
@henderiw I've managed to run your github.com/henderiw-nephio/pkg-examples -> ipam-fn code,
and for me the SetStatus and SetFromTypedObject versions ouput the exact same result (character-by character).
What is your error?